src/npc.cpp

```diff
@@ -1923,11 +1933,13 @@ void npc::die( Creature *nkiller )
     dead = true;
     Character::die( nkiller );
 
-    if( g->u.sees( *this ) ) {
+    if( g->u.sees( *this ) && !is_hallucination() ) {
```

Comment:
I'd prefer to see:
```
    if( is_hallucination() ) {
        if( g->u.sees( *this ) ) {
            add_msg( _( "%s disappears." ), name );
        }
        return;
    }
```
and then get rid of all the other special logic in this function.